api/v2/routes_payment.go

```diff
@@ -290,6 +293,57 @@ func (api *API) CreateDashPayment(c *gin.Context) {
 	Respond(c, http.StatusOK, gin.H{"response": p})
 }
 
+func (api *API) stripeCharge(c *gin.Context) {
+	username, err := GetAuthenticatedUserFromContext(c)
+	if err != nil {
+		api.LogError(c, err, eh.NoAPITokenError)(http.StatusBadRequest)
+		return
+	}
+	forms, missingField := api.extractPostForms(c, "stripe_token", "stripe_email", "value_in_cents")
+	if missingField != "" {
+		FailWithMissingField(c, missingField)
+		return
+	}
+	valueInCentsInt, err := strconv.ParseInt(forms["value_in_cents"], 10, 64)
+	if err != nil {
+		Fail(c, err)
+		return
+	}
+	valueInCentsFloat, err := strconv.ParseFloat(forms["value_in_cents"], 64)
+	if err != nil {
+		Fail(c, err)
+	}
+	// set the secret key after validation
+	stripe.Key = api.cfg.Stripe.SecretKey
+	// initialize credit card charge parameters
+	chargeParams := &stripe.ChargeParams{
+		Amount:      stripe.Int64(valueInCentsInt),
+		Currency:    stripe.String(string(stripe.CurrencyUSD)),
+		Description: stripe.String("temporal credit purchase"),
+		// StatementDescriptor is what appears in their credit card billing report
+		StatementDescriptor: stripe.String("credit purchase"),
+		// email the receipt goes to
+		ReceiptEmail: stripe.String(forms["stripe_email"]),
+	}
+	// set the token generated by stripe
+	chargeParams.SetSource(forms["stripe_token"])
+	// metadata can be used for better Radar validation
+	chargeParams.AddMetadata("order_type", "temporal.credits")
+	ch, err := charge.New(chargeParams)
+	if err != nil {
+		api.LogError(c, err, err.Error())(http.StatusBadRequest)
+		return
+	}
+	api.l.Infow("payment complete", "payment.method", "stripe", "user", username, "charge", ch)
+	// add credits to use
+	if _, err := api.um.AddCredits(username, valueInCentsFloat/100); err != nil {
+		api.LogError(c, err, "failed to grant credits")(http.StatusBadRequest)
```

Comment:
is this really a bad request?

Reply:
Good point I'll change to internal server error